### PR TITLE
Migration of all blobs to new Riak cluster

### PR DESCRIPTION
**Pull request by millerdev**
I continued with the existing pattern of blob db migrations. We'll need to add two new localsettings:

```py
BLOB_DB_MIGRATING_FROM_S3_TO_S3 = True
OLD_S3_BLOB_DB_SETTINGS = {...}
```

The test is slow and ended up with way too much boilerplate for setup 😞 

@snopoke @javierwilson

### CODE REVIEW

File: corehq/blobs/migrate.py [hunk 1]
```diff
@@ -378,14 +403,100 @@ def process_object(self, object):
...
+
+    This is currently built on the assumtion that the SQL model only
+    references a single blob, and the blob name is not used.
+    """
+
+    def __init__(self, obj, blob_id, bucket):
+        self.obj = obj
+        self.blobs = {"name-ignored": BlobMeta(id=blob_id)}
+        self.external_blobs = self.blobs
+        self._blobdb_bucket = lambda:bucket
```
Comment (dimagimon) [inline corehq/blobs/migrate.py]: E231 missing whitespace after ':'

File: corehq/blobs/migrate.py [hunk 2]
```diff
@@ -411,6 +522,50 @@ def _get_progress_logger(self):
         return CouchProcessorProgressLogger(self.doc_types)
 
 
+class SqlMigrator(Migrator):
+
+    def __init__(self, slug, reindexer, doc_migrator_class):
+        types = [reindexer.model_class]
+        assert not hasattr(types[0], "get_db"), types[0] # not a couch model
```
Comment (dimagimon) [inline corehq/blobs/migrate.py]: E261 at least two spaces before inline comment

File: corehq/blobs/migrate.py [hunk 3]
```diff
@@ -411,6 +522,50 @@ def _get_progress_logger(self):
...
+    def _get_document_provider(self):
+        return SqlDocumentProvider(self.iteration_key, self.reindexer)
+
+
+class MultiDbMigrator(object):
+
+    def __init__(self, slug, couch_types, sql_reindexers, doc_migrator_class):
+        self.slug = slug
+        self.migrators = migrators = []
+        def db_key(doc_type):
```
Comment (dimagimon) [inline corehq/blobs/migrate.py]: E301 expected 1 blank line, found 0

File: corehq/blobs/tests/test_migrate.py [hunk 4]
```diff
@@ -434,31 +440,147 @@ def modify(item):
...
 
-        fsdb = TemporaryFilesystemBlobDB()
-        assert get_blob_db() is fsdb, (get_blob_db(), fsdb)
-        self.migrate_docs = docs = []
-        for i in range(self.test_size):
-            doc = SavedBasicExport(configuration=_mk_config("config-%s" % i))
-            doc.save()
-            doc.set_payload(("content %s" % i).encode('utf-8'))
-            docs.append(doc)
+        lost_db = TemporaryFilesystemBlobDB() # must be created before S3 dbs
```
Comment (dimagimon) [inline corehq/blobs/tests/test_migrate.py]: E261 at least two spaces before inline comment

File: corehq/blobs/migrate.py [hunk 5]
```diff
@@ -411,6 +522,50 @@ def _get_progress_logger(self):
...
+        def filen(n):
+            return None if filename is None else "{}.{}".format(filename, n)
+        migrated = 0
+        skipped = 0
+        for n, item in enumerate(self.migrators):
+            one_migrated, one_skipped = item.migrate(filen(n), *args, **kw)
+            migrated += one_migrated
+            skipped += one_skipped
+            print("\n")
+        return migrated, skipped
```
Comment (millerdev) [inline corehq/blobs/migrate.py]: Note to self: need to wait to write `BlobMigrationState` until all sub-migrations are complete. It's currently written after each sub-migration, which works but is not correct.

**Comment by millerdev** [comment]
@esoergel code buddy

**Comment by esoergel** [comment]
Failure looks real:
```
======================================================================
ERROR: test_iteration (corehq.util.tests.test_doc_processor.CaseResumableSqlModelIteratorTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/commcare-hq-ro/corehq/util/tests/test_doc_processor.py", line 288, in setUp
    super(CaseResumableSqlModelIteratorTest, self).base_setUp()
  File "/mnt/commcare-hq-ro/corehq/util/tests/test_doc_processor.py", line 184, in base_setUp
    self.itr = self.get_iterator()
  File "/mnt/commcare-hq-ro/corehq/util/tests/test_doc_processor.py", line 191, in get_iterator
    return resumable_sql_model_iterator(self.iteration_key, reindex_accessor, chunk_size)
  File "/mnt/commcare-hq-ro/corehq/util/doc_processor/sql.py", line 54, in resumable_sql_model_iterator
    if reindex_accessor.is_sharded():
  File "/mnt/commcare-hq-ro/corehq/form_processor/backends/sql/dbaccessors.py", line 152, in is_sharded
    return isinstance(cls.model_class.objects, RestrictedManager)
AttributeError: 'property' object has no attribute 'objects'
```

File: corehq/blobs/tests/test_migrate.py [hunk 6]
```diff
@@ -434,31 +440,147 @@ def modify(item):
...
+                    lost.id,
+                    lost_blob_id,
+                    rex.blob_helper({"_obj_not_json": lost})._blobdb_bucket(),
+                ))
+
+        self.test_size = len(self.couch_docs) + len(self.sql_docs)
+        db2 = TemporaryS3BlobDB(config)
+        self.db = TemporaryMigratingBlobDB(db2, db1)
         assert get_blob_db() is self.db, (get_blob_db(), self.db)
         BaseMigrationTest.discard_migration_state(self.slug)
```
Comment (esoergel) [inline corehq/blobs/tests/test_migrate.py]: Is this all necessary to happen on `setUp`?  Can anything be in `setUpClass` instead?

nit: this setup function is quite long and difficult to follow, it'd be nice for it to be broken into separate method calls.

Comment (esoergel) [inline corehq/blobs/tests/test_migrate.py]: Oh, I guess it's all one test method, so the `setUp` vs `setUpClass` distinction doesn't really matter.

**Review by esoergel — COMMENTED** [review]
Has the word "blob" started to look weird for you yet?

File: corehq/blobs/tests/test_migrate.py (hunk 6 above)
Comment (millerdev) [inline corehq/blobs/tests/test_migrate.py]: Most of it could be moved `setUpClass`, but since it's  only one test...

Agree on the long and hard to follow comment, and I'm not happy with how much code was needed to setup for this test.
